neps/nep-0393.md

```diff
@@ -0,0 +1,516 @@
+
+    /// Returns total amount of tokens issued by `ctr` SBT contract, including expired tokens.
+    /// Depending on the implementation, if a revoke removes a token, it then is should not be
+    /// included in the supply.
+    fn sbt_supply(&self, ctr: AccountId) -> u64;
+
+    /// Returns total amount of tokens of given class minted by `ctr`. See `sbt_supply` for
+    /// information about revoked tokens.
+    fn sbt_supply_by_class(&self, ctr: AccountId, class: ClassId) -> u64;
+
+    /// Returns total supply of SBTs for a given owner. See `sbt_supply` for information about
+    /// revoked tokens.
+    /// If class is specified, returns only owner supply of the given class -- must be 0 or 1.
+    fn sbt_supply_by_owner(
+        &self,
+        account: AccountId,
+        ctr: AccountId,
+        class: Option<ClassId>,
+    ) -> u64;
+
+    /// Query sbt tokens issued by a given contract.
+    /// If `from_token` is not specified, then `from_token` should be assumed
+    /// to be the first valid token id.
+    fn sbt_tokens(&self, ctr: AccountId, from_token: Option<u64>, limit: Option<u32>)
+        -> Vec<Token>;
+
+    /// Query SBT tokens by owner
+    /// If `from_class` is not specified, then `from_class` should be assumed to be the first
+    /// valid class id.
+    /// Returns list of pairs: `(Contract address, list of token IDs)`.
+    fn sbt_tokens_by_owner(
+        &self,
+        account: AccountId,
+        ctr: Option<AccountId>,
+        from_class: Option<u64>,
+        limit: Option<u32>,
+    ) -> Vec<(AccountId, Vec<OwnedToken>)>;
+
+    /// checks if an `account` was banned by the registry.
+    fn is_banned(&self, account: AccountId) -> bool;
+
+    /*************
+     * Transactions
+     *************/
+
+    /// Creates a new, unique token and assigns it to the `receiver`.
+    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
+    /// Each TokenMetadata must have non zero `class`.
+    /// Must be called by an SBT contract.
+    /// Must emit `Mint` event.
+    /// Must provide enough NEAR to cover registry storage cost.
+    // #[payable]
+    fn sbt_mint(&mut self, token_spec: Vec<(AccountId, Vec<TokenMetadata>)>) -> Vec<TokenId>;
+
+    /// sbt_recover reassigns all tokens from the old owner to a new owner,
+    /// and registers `old_owner` to a burned addresses registry.
+    /// Must be called by an SBT contract.
+    /// Must emit `Recover` event.
+    /// Must be called by an operator.
+    /// Must provide enough NEAR to cover registry storage cost.
+    /// Requires attaching enough tokens to cover the storage growth.
+    // #[payable]
+    fn sbt_recover(&mut self, from: AccountId, to: AccountId);
+
+    /// sbt_renew will update the expire time of provided tokens.
+    /// `expires_at` is a unix timestamp (in miliseconds).
+    /// Must be called by an SBT contract.
+    /// Must emit `Renew` event.
+    fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64);
```

Comment:
This seems very use-case specific, and I don't think it belongs in the standard.

Reply:
Why? Remember that this is called by the issuer, not the user. So all the specificity is handled by the issuer contract, and the issuer will only be able to send a request to update the registry. Registry doesn't need to know that custom circumstances of the issuer.

Reply:
You see, this again highlights the problem with the registry concept. If the registry doesn't support some feature, it's very difficult for the issuer to do it themselves, without becoming a registry themselves. Because I'd expect most SBTs wouldn't be renewable, but it cannot be an application specific interface, it has to be on the standard itself.

With any new primitive, it's not easy to predict the actual use cases. If the core interface doesn't cover that use case, it'll either result in new NEPs, or you end up with interface methods noone uses.

Reply:
I wonder whether it makes sense to define `sbt_renew` as an (optional) extension to the core standard?

There are a few other capabilities in the SBT sample implementation that are not part of spec. So I don't see an issue in having a few capabilities like this one included in the spec. I will prefer however to keep the required capabilities to a minimum to make adoption easier.